File: cses/06_tree_algorithms/2081_fixed-length-paths-ii/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); pos += 1
    k1 = int(data[pos]); pos += 1
    k2 = int(data[pos]); pos += 1

    # Danh sách kề của cây
    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        a = int(data[pos]); b = int(data[pos + 1]); pos += 2
        adj[a].append(b)
        adj[b].append(a)

    removed = bytearray(n + 1)   # đỉnh đã bị chọn làm centroid và xóa
    subsize = [0] * (n + 1)      # kích thước cây con trong component hiện tại
    par = [0] * (n + 1)          # cha khi duyệt component để tính size
    cpar = [0] * (n + 1)         # cha khi duyệt depth từ centroid
    depthv = [0] * (n + 1)       # depth (số cạnh) từ centroid tới mỗi đỉnh

    k1m1 = k1 - 1

    def count_leq(arr, X):
        # Số cặp không thứ tự (i < j) có arr[i] + arr[j] <= X, arr đã sắp tăng.
        # Two-pointer O(len).
        l = 0
        r = len(arr) - 1
        c = 0
        while l < r:
            if arr[l] + arr[r] <= X:
                c += r - l
                l += 1
            else:
                r -= 1
        return c

    ans = 0
    stack = [1]   # các component chờ xử lý (đại diện bởi một đỉnh bất kỳ)

    while stack:
        start = stack.pop()
        if removed[start]:
            continue

        # --- Duyệt component (iterative) và tính subtree size khi root tại start ---
        comp = []
        par[start] = 0
        st = [start]
        while st:
            u = st.pop()
            comp.append(u)
            pu = par[u]
            for v in adj[u]:
                if not removed[v] and v != pu:
                    par[v] = u
                    st.append(v)
        # Duyệt ngược thứ tự phát hiện để cộng size từ lá lên gốc
        for u in reversed(comp):
            s = 1
            pu = par[u]
            for v in adj[u]:
                if not removed[v] and v != pu:
                    s += subsize[v]
            subsize[u] = s
        total = len(comp)

        # --- Tìm centroid: đi về phía cây con có kích thước > total/2 ---
        u = start
        p = 0
        while True:
            best = -1
            for v in adj[u]:
                if removed[v] or v == p:
                    continue
                if 2 * subsize[v] > total:
                    best = v
                    break
            if best == -1:
                break
            p = u
            u = best
        c = u

        # --- Đóng góp của các path đi qua centroid c (inclusion-exclusion) ---
        # D_all = centroid (depth 0) + mọi node trong component (depth tính từ c).
        # contrib = count_range(D_all) - sum_child count_range(child).
        D_all = [0]
        contrib = 0
        for nb in adj[c]:
            if removed[nb]:
                continue
            # DFS iterative trong một cây con của c để lấy mảng depth
            cpar[nb] = c
            depthv[nb] = 1
            st2 = [nb]
            child = []
            capp = child.append
            while st2:
                x = st2.pop()
                dx = depthv[x]
                capp(dx)
                px = cpar[x]
                for v in adj[x]:
                    if not removed[v] and v != px:
                        cpar[v] = x
                        depthv[v] = dx + 1
                        st2.append(v)
            child.sort()
            # Trừ đi các cặp cùng một cây con (đường đi thật giữa chúng không qua c)
            if len(child) >= 2:
                contrib -= count_leq(child, k2) - count_leq(child, k1m1)
            D_all += child
        # count_range trên toàn bộ D_all rồi cộng vào đáp số
        D_all.sort()
        contrib += count_leq(D_all, k2) - count_leq(D_all, k1m1)
        ans += contrib

        # --- Xóa centroid, đẩy các component con vào stack để đệ quy ---
        removed[c] = 1
        for nb in adj[c]:
            if not removed[nb]:
                stack.append(nb)

    sys.stdout.buffer.write((str(ans) + "\n").encode())


main()

File: cses/06_tree_algorithms/2081_fixed-length-paths-ii/solution.cpp
#include <algorithm>
#include <iostream>
#include <tuple>
#include <utility>
#include <vector>

// Một cây con (branch) của centroid: danh sách depth của các đỉnh và depth lớn nhất
struct DepthBranch {
    std::vector<int> depths;
    int maximum_depth = 0;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, minimum_length, maximum_length;
    std::cin >> n >> minimum_length >> maximum_length;

    // Danh sách kề của cây
    std::vector<std::vector<int>> graph(n + 1);
    for (int edge = 0; edge < n - 1; ++edge) {
        int first, second;
        std::cin >> first >> second;
        graph[first].push_back(second);
        graph[second].push_back(first);
    }

    std::vector<char> removed(n + 1, false);      // đỉnh đã bị chọn làm centroid và xóa
    std::vector<int> parent(n + 1, 0);            // cha khi duyệt component để tính size
    std::vector<int> subtree_size(n + 1, 0);      // kích thước cây con trong component
    std::vector<int> frequency(maximum_length + 1, 0);  // đếm số đỉnh theo từng depth
    std::vector<int> prefix(maximum_length + 1, 0);     // tổng tiền tố của frequency
    std::vector<int> component;
    component.reserve(n);
    std::vector<int> component_stack{1};          // các component chờ xử lý
    long long answer = 0;

    while (!component_stack.empty()) {
        const int start = component_stack.back();
        component_stack.pop_back();
        if (removed[start]) {
            continue;
        }

        // --- Duyệt component (iterative) khi root tại start ---
        component.clear();
        std::vector<int> traversal{start};
        parent[start] = 0;
        while (!traversal.empty()) {
            const int node = traversal.back();
            traversal.pop_back();
            component.push_back(node);
            for (const int next : graph[node]) {
                if (removed[next] || next == parent[node]) {
                    continue;
                }
                parent[next] = node;
                traversal.push_back(next);
            }
        }

        // Tính subtree size: duyệt ngược thứ tự phát hiện để cộng từ lá lên gốc
        for (const int node : component) {
            subtree_size[node] = 1;
        }
        for (auto iterator = component.rbegin();
             iterator != component.rend(); ++iterator) {
            const int node = *iterator;
            if (parent[node] != 0) {
                subtree_size[parent[node]] += subtree_size[node];
            }
        }

        // --- Tìm centroid: đỉnh mà mọi phần tách ra đều <= total/2 ---
        const int total_size = static_cast<int>(component.size());
        int centroid = start;
        for (const int node : component) {
            int largest_part = total_size - subtree_size[node];
            for (const int next : graph[node]) {
                if (!removed[next] && parent[next] == node) {
                    largest_part = std::max(
                        largest_part, subtree_size[next]
                    );
                }
            }
            if (largest_part * 2 <= total_size) {
                centroid = node;
                break;
            }
        }

        // --- Với mỗi cây con của centroid, DFS iterative lấy mảng depth ---
        // Chỉ giữ các depth <= maximum_length vì phần dài hơn không thể tạo path hợp lệ
        std::vector<DepthBranch> branches;
        branches.reserve(graph[centroid].size());
        for (const int neighbor : graph[centroid]) {
            if (removed[neighbor]) {
                continue;
            }

            DepthBranch branch;
            std::vector<std::tuple<int, int, int>> depth_stack;
            depth_stack.emplace_back(neighbor, centroid, 1);
            while (!depth_stack.empty()) {
                const auto [node, ancestor, depth] = depth_stack.back();
                depth_stack.pop_back();
                if (depth > maximum_length) {
                    continue;
                }

                branch.depths.push_back(depth);
                branch.maximum_depth = std::max(
                    branch.maximum_depth, depth
                );
                for (const int next : graph[node]) {
                    if (!removed[next] && next != ancestor) {
                        depth_stack.emplace_back(next, node, depth + 1);
                    }
                }
            }
            if (!branch.depths.empty()) {
                branches.push_back(std::move(branch));
            }
        }

        // Mỗi cặp được đếm khi xử lý cây con thứ hai của nó. Sắp các cây con theo
        // depth lớn nhất giúp việc dựng lại mảng prefix-frequency chỉ tốn tuyến
        // tính theo tổng kích thước các cây con tại centroid này.
        std::sort(
            branches.begin(), branches.end(),
            [](const DepthBranch& first, const DepthBranch& second) {
                return first.maximum_depth < second.maximum_depth;
            }
        );

        frequency[0] = 1;  // depth 0 là chính centroid, để đếm path một đầu là centroid
        int current_maximum = 0;
        for (const DepthBranch& branch : branches) {
            // Dựng prefix-sum của frequency (các cây con đã xét trước đó)
            prefix[0] = frequency[0];
            for (int depth = 1; depth <= current_maximum; ++depth) {
                prefix[depth] = prefix[depth - 1] + frequency[depth];
            }

            // Ghép mỗi đỉnh của cây con hiện tại với các đỉnh cây con trước đó sao
            // cho tổng depth nằm trong [minimum_length, maximum_length]
            for (const int depth : branch.depths) {
                const int low = std::max(0, minimum_length - depth);
                const int high = std::min(
                    current_maximum, maximum_length - depth
                );
                if (low <= high) {
                    answer += prefix[high];
                    if (low > 0) {
                        answer -= prefix[low - 1];
                    }
                }
            }

            // Thêm depth của cây con hiện tại vào frequency cho các cây con sau
            for (const int depth : branch.depths) {
                ++frequency[depth];
            }
            current_maximum = std::max(
                current_maximum, branch.maximum_depth
            );
        }
        // Reset frequency về 0 cho centroid tiếp theo
        std::fill(
            frequency.begin(), frequency.begin() + current_maximum + 1, 0
        );

        // --- Xóa centroid, đẩy các component con vào stack để đệ quy ---
        removed[centroid] = true;
        for (const int next : graph[centroid]) {
            if (!removed[next]) {
                component_stack.push_back(next);
            }
        }
    }

    std::cout << answer << '\n';
    return 0;
}
